src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -6549,14 +6554,29 @@ public override IOperation VisitTranslatedQuery(ITranslatedQueryOperation operat
 
         public override IOperation VisitConstantPattern(IConstantPatternOperation operation, int? captureIdForResult)
         {
-            return new ConstantPatternOperation(Visit(operation.Value), semanticModel: null,
+            return new ConstantPatternOperation(operation.InputType, Visit(operation.Value), semanticModel: null,
                 syntax: operation.Syntax, isImplicit: IsImplicit(operation));
         }
 
         public override IOperation VisitDeclarationPattern(IDeclarationPatternOperation operation, int? captureIdForResult)
         {
-            return new DeclarationPatternOperation(operation.DeclaredSymbol, semanticModel: null,
-                operation.Syntax, operation.Type, operation.ConstantValue, IsImplicit(operation));
+            return new DeclarationPatternOperation(
+                inputType: operation.InputType, matchedType: operation.MatchedType, operation.DeclaredSymbol, operation.MatchesNull, semanticModel: null,
+                operation.Syntax, IsImplicit(operation));
+        }
+
+        public override IOperation VisitRecursivePattern(IRecursivePatternOperation operation, int? argument)
+        {
+            return new RecursivePatternOperation(
+                inputType: operation.InputType,
+                matchedType: operation.MatchedType,
+                operation.DeconstructSymbol,
+                operation.DeconstructionSubpatterns.SelectAsArray(p => (IPatternOperation)Visit(p)),
```

Comment:
>operation.DeconstructionSubpatterns.SelectAsArray(p => (IPatternOperation)Visit(p)), [](start = 16, length = 84)

The current rewrite would be safe only if we can be certain that none of the constituent IOperation nodes can contain a control flow. However, it looks like a constant pattern accepts a constant in the following form ```(true ? 1 : 2)```. What would happen with the flow graph when such constant expression is used inside recursive pattern? Should also think about other possible edge cases, including error cases.

Reply:
It looks like VisitIsPattern handles possible control flow in the pattern, perhaps we should do the same here.

---
In reply to: [244200140](https://github.com/dotnet/roslyn/pull/32018#discussion_r244200140) [](ancestors = 244200140)

Reply:
This is a pattern, not an expression. The recursive pattern has exactly as much support for control-flow within it as the current implementation of `VisitConstantPattern`.  I will add tests to demonstrate.

---
In reply to: [244200861](https://github.com/dotnet/roslyn/pull/32018#discussion_r244200861) [](ancestors = 244200861,244200140)

Reply:
> This is a pattern, not an expression. 

Its constituent parts contain expressions

> The recursive pattern has exactly as much support for control-flow within it as the current implementation of VisitConstantPattern.

The VisitConstantPattern doesn't need special handling because it has only one sup-operation. It is a bad example.

---
In reply to: [244211373](https://github.com/dotnet/roslyn/pull/32018#discussion_r244211373) [](ancestors = 244211373,244200861,244200140)

Reply:
This thread is marked as resolved, but I do not see any implementation changes or new tests demonstrating absence of problems added. The kind of test that would be interesting is a scenario with a control flow somewhere at the end of the recursive pattern, in the last sub-pattern, etc. 

---
In reply to: [244211893](https://github.com/dotnet/roslyn/pull/32018#discussion_r244211893) [](ancestors = 244211893,244211373,244200861,244200140)

Reply:
I think the added test (IsPattern_ControlFlowInPattern_02) isn't a good representative of possible shortcomings of the current rewrite. It is not blocking for this PR, but please open an issue to follow-up on this subject.

---
In reply to: [244428535](https://github.com/dotnet/roslyn/pull/32018#discussion_r244428535) [](ancestors = 244428535,244211893,244211373,244200861,244200140)